api/src/context.rs

```diff
@@ -666,33 +679,97 @@ impl Context {
-            gas_estimation.median_gas_price = if gas_prices.is_empty() {
-                // Pull the min gas price
-                let gas_schedule = self.get_gas_schedule(ledger_info)?;
-                gas_estimation.min_gas_price = gas_schedule.txn.min_price_per_gas_unit.into();
-                gas_estimation.min_gas_price
+            gas_estimation.last_updated_epoch = Some(last_updated_epoch);
+            let min_gas_price = gas_schedule.txn.min_price_per_gas_unit.into();
+            let max_gas_price = gas_schedule.txn.max_price_per_gas_unit.into();
+
+            let observed_gas_median = if gas_prices.is_empty() {
+                // This will make it always pick the minimum gas price
+                0
             } else {
+                // Sort and take the median of the gas prices
                 gas_prices.sort();
+
+                // Median is the center if it's odd, average of the two middle if even
                 let mid = gas_prices.len() / 2;
-                gas_prices[mid]
+                if gas_prices.len() % 2 == 0 {
+                    let lower = gas_prices.get(mid).unwrap();
+                    let upper = gas_prices.get(mid + 1).unwrap();
+
+                    upper.saturating_add(*lower).saturating_div(2)
+                } else {
+                    *gas_prices.get(mid).unwrap()
+                }
             };
 
-            Ok(gas_estimation.median_gas_price)
+            // Ensure the price is within the min/max bounds
+            gas_estimation.median_gas_price = std::cmp::min(
+                max_gas_price,
+                std::cmp::max(min_gas_price, observed_gas_median),
+            );
+
+            // Handle prices for "prioritized" and "deprioritized"
+            // There must be at least 1 buckets
+            assert!(!self.node_config.mempool.broadcast_buckets.is_empty());
+            let median_gas_price = gas_estimation.median_gas_price;
+
+            // TODO: Buckets are assumed to be ordered
+            let bucket_index = self
+                .node_config
+                .mempool
+                .broadcast_buckets
+                .iter()
+                .sorted()
+                .enumerate()
+                .find_map(|(index, current_bucket)| {
+                    if median_gas_price >= *current_bucket {
+                        Some(index)
+                    } else {
+                        None
+                    }
+                })
+                .unwrap_or_default();
+
+            // Previous bucket could be the same as the minimum price (if too low)
+            let previous_bucket_price = self
+                .node_config
+                .mempool
+                .broadcast_buckets
+                .get(bucket_index.saturating_sub(1))
+                .copied()
+                .unwrap_or(min_gas_price);
+            // Next bucket could be the same as the current gas price, but we increase by 1 for the estimation
```

Comment:
Why do we need to add 1? The bucket already has it (e.g. 151, 301, etc. https://github.com/aptos-labs/aptos-core/blob/main/config/src/config/mempool_config.rs#L47)